### PR TITLE
MultiCheckpointWriter Bug Fix

### PR DESCRIPTION
**Pull request by annym**
Whenever the last address of a stream is a hole (right before the checkpointer is triggered)
We build the checkpoint start log address based on the wrong start of the stream,
this is because we discard holes and hence set the start to the previous DATA entry.
This will diverge from the state kept by the sequencer for this stream on trims (trim mark).

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 1]
```diff
@@ -184,6 +200,15 @@ private void appendCheckpointOnSnapshot(Token snapshotTimestamp) {
         } finally {
             rt.getObjectsView().TXEnd();
         }
+
+        return snapshotTimestamp;
+    }
+
+    private Token forceHole() {
+        Token writeToken = rt.getSequencerView().next(streamId).getToken();
+        LogData hole = LogData.getHole(writeToken);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Its better to use `LogData::useToken` instead of `LogData.getHole`

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 2]
```diff
@@ -172,7 +188,7 @@ private void appendCheckpointOnSnapshot(Token snapshotTimestamp) {
             log.info("appendCheckpoint: Started checkpoint for {} at snapshot {}", streamId, snapshotTimestamp);
             ICorfuSMR<T> corfuObject = (ICorfuSMR<T>) this.map;
             Set<Map.Entry> entries = this.map.entrySet();
-            long vloVersion = corfuObject.getCorfuSMRProxy().getVersion();
+            long vloVersion = snapshotTimestamp.getSequence();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: This needs a comment about why its reasonable to "change" the vlo version

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java [hunk 3]
```diff
@@ -1002,4 +1002,78 @@ public void testStreamRebuiltWithHoleAsFirstEntryAfterTrimNoCP() throws Exceptio
...
+        // Write 10 Entries to streamA
+        for (int i = 0; i < numEntries; i++) {
+            mA.put(String.valueOf(i), i);
+        }
+
+        // Force a hole as the last update to streamA
+        Token token = runtime.getSequencerView().next(CorfuRuntime.getStreamID(streamA)).getToken();
+        LogData hole = LogData.getHole(token);
+        runtime.getLayoutView().getRuntimeLayout()
+                .getLogUnitClient("tcp://localhost:9000")
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: You can use a NodeLocator here.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java [hunk 4]
```diff
@@ -1002,4 +1002,78 @@ public void testStreamRebuiltWithHoleAsFirstEntryAfterTrimNoCP() throws Exceptio
...
+        CheckpointWriter cpwB = new CheckpointWriter(runtime, CorfuRuntime.getStreamID(streamB),
+                "checkpointer-Test", mB);
+        cpwB.appendCheckpoint();
+
+        // Add an update to streamA after checkpoint
+        mA.put(String.valueOf(numEntries), numEntries);
+
+        // Trim the log at B's CPToken
+        runtime.getAddressSpaceView().prefixTrim(cpTokenA);
+
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Restart, or flush the server caches after the trim

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java [hunk 5]
```diff
@@ -1002,4 +1002,78 @@ public void testStreamRebuiltWithHoleAsFirstEntryAfterTrimNoCP() throws Exceptio
...
+        // Instantiate streamA from new Runtime, so stream is rebuilt
+        CorfuRuntime rt2 = new CorfuRuntime(DEFAULT_ENDPOINT).connect();
+        Map<String, Integer> mA2 = createMap(rt2, streamA);
+
+        // By accessing the map, we ensure we are able to load from the checkpoint.
+        try {
+            rt2.getObjectsView().TXBuild()
+                    .type(TransactionType.OPTIMISTIC)
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Can you also check the data ?

File: test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java [hunk 6]
```diff
@@ -567,7 +577,7 @@ public void MultiCheckpointWriterTestInner(boolean consecutiveCkpoints) throws E
         mcw2.addMap((SMRMap) mA);
         mcw2.addMap((SMRMap) mB);
         long firstGlobalAddress2 = mcw2.appendCheckpoints(r, author).getSequence();
-        assertThat(firstGlobalAddress2).isGreaterThan(firstGlobalAddress1);
+        assertThat(firstGlobalAddress2).isGreaterThanOrEqualTo(firstGlobalAddress1);
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java]: Why isn't this a more strict check ? (i.e. just equality)

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java [hunk 7]
```diff
@@ -1002,4 +1002,78 @@ public void testStreamRebuiltWithHoleAsFirstEntryAfterTrimNoCP() throws Exceptio
...
+    /**
+     * Test rebuilding a stream from a new runtime, whenever the last address before the checkpoint
+     * was a hole.
+     *
+     * This case is interesting to test as addresses that become holes are discarded by the streamView,
+     * while sequencer's are agnostic of this info, hence, take it into account for trim mark computation.
+     */
+    @Test
+    public void testCheckpointWhenLastAddressIsHole() throws Exception {
+        final int numEntries = 10;
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Do we need a test that creates an empty checkpoint and restarts the server to verify that the sequencer bootstrap state is correct?

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 8]
```diff
@@ -148,19 +146,37 @@ public Token appendCheckpoint() {
...
+        // Note: we could simply not checkpoint empty streams with the introduction
+        // of address maps, but for now, we'll keep it consistent with old backpointer
+        // strategy, which required checkpoints for empty streams to avoid TrimmedExceptions
+        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+            snapshotTimestamp = forceHole();
+        }
+
+        try {
+            rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
+        } catch (TrimmedException te) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Do you really need this? The idea of writing a whole is to mark a slot in the address space to prevent the sequencer regressing, but if its trimmed, then we can't regress from a trimmed address, right?

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 1 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 9]
```diff
@@ -148,19 +146,37 @@ public Token appendCheckpoint() {
...
+        // If snapshotTimestamp (stream tail) is a negative value
+        // we need to enforce a hole to ensure the start log address
+        // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
+        // Having a negative value will prevent trimming from happening.
+
+        // Note: we could simply not checkpoint empty streams with the introduction
+        // of address maps, but for now, we'll keep it consistent with old backpointer
+        // strategy, which required checkpoints for empty streams to avoid TrimmedExceptions
+        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+            snapshotTimestamp = forceHole();
```
Comment (lixinchengdu) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Is it better to define another type of LogData to represent dummy data but has backpointers? If we still want checkpoint to be compatible with back-pointer mechanism, we should avoid holes if possible.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 2 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Done.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 3 above)
Comment (annym) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Done.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 4 above)
Comment (annym) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Done.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 5 above)
Comment (annym) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Done.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 7 above)
Comment (annym) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: Sure, this is a good idea, cause the test we have does not exercise reboot. Adding it.

File: test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java (hunk 6 above)
Comment (annym) [reply test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java]: Because this method serves the purpose of two different tests (L.571) so in the case of consecutive checkpoints it will be greater cause the stream tail will change, if not, it will be equal cause the stream tail won't change.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 8 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: We do need it, because consider the case where the stream has not been updated for several checkpoint cycles, therefore the stream tail has not changed, but after a couple of cycles let's say it is trimmed and entirely subsumed by a checkpoint. This read will not serve the purpose of hole fill in that case and will throw a TrimmedException, cause it is no longer present.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 9 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: I don't really think we need that, anyways our goal is to completely move away from back pointers. And address maps does not have this issue.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 8 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: I misunderstood this comment, thinking it was related to catching the TrimmedException but I believe it was actually related to forcing a hole in these cases. Which I agree with, there is no case in filling a hole as there cannot be regression in this case. Removing this!

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 10]
```diff
@@ -148,19 +147,37 @@ public Token appendCheckpoint() {
      */
     @VisibleForTesting
     public Token appendCheckpoint(Token snapshotTimestamp) {
-        appendCheckpointOnSnapshot(snapshotTimestamp);
-        return snapshotTimestamp;
+        return appendCheckpointOnSnapshot(snapshotTimestamp);
```
[the remark this answers is not in the document]
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: appendCheckpoint is public and only visible for testing.. this other is private.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 11]
```diff
@@ -130,14 +132,11 @@ public CheckpointWriter(CorfuRuntime rt, UUID streamId, String author, T map) {
      * @return Token at which the snapshot for this checkpoint was taken.
      */
     public Token appendCheckpoint() {
-        // Queries the sequencer for the global log tail. We then read the global log tail to
+        // Queries the sequencer for the stream tail. We then read the stream tail to
         // persist the entry on the log unit in turn preventing from a new sequencer from
```
[the remark this answers is not in the document]
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: @annym  We used to have two different methods because they have different return type, but now they are exactly same.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 2 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Instead of reassigning the vloVersion variable, you can check if long vloVersion == corfuObject.getCorfuSMRProxy().getVersion(); and maybe log something ? do we even care?
